pycc/ccwfn.py

```diff
@@ -333,10 +336,42 @@ def residuals(self, F, t1, t2):
 
         r1 = self.r_T1(o, v, F, ERI, L, t1, t2, Fae, Fme, Fmi)
         r2 = self.r_T2(o, v, F, ERI, L, t1, t2, Fae, Fme, Fmi, Wmnij, Wmbej, Wmbje, Zmbij)
-       
+             
         if isinstance(Fae, torch.Tensor):
-            del Fae, Fmi, Fme, Wmnij, Wmbej, Wmbje, Zmbij 
-
+            del Fae, Fmi, Wmnij, Wmbej, Wmbje, Zmbij 
+        
+        if self.model == 'CC3':
+            # Build intermediates
+            Wmnij_cc3 = self.build_cc3_Wmnij(o, v, ERI, t1)
+            Wmbij_cc3 = self.build_cc3_Wmbij(o, v, ERI, t1, Wmnij_cc3)
+            Wmnie_cc3 = self.build_cc3_Wmnie(o, v, ERI, t1)
+            Wamef_cc3 = self.build_cc3_Wamef(o, v, ERI, t1)
+            Wabei_cc3 = self.build_cc3_Wabei(o, v, ERI, t1)
+              
+            # Loop over sets of ijk
+            for i in range(no):
+                for j in range(no):
+                    for k in range(no):
+                        # t_ijk                        
+                        t3 = cctriples.t3c_ijk(self, o, v, i, j, k, t2, Wabei_cc3, Wmbij_cc3, F, WithDenom=True)
+                        # t_ijkabc -> Sia
+                        r1[i] += self.sigma_T1(v, j, k, L, t3)                       
+                        # t_ijkabc -> Sijab
+                        tmp = self.sigma_T2_A(k, Fme, t3)
+                        r2[i,j] += tmp
+                        r2[j,i] += tmp.swapaxes(0,1)                      
+                        # t_ijkabc -> Sijad
+                        tmp = self.sigma_T2_B(k, Wamef_cc3, t3)
+                        r2[i,j] += tmp
+                        r2[j,i] += tmp.swapaxes(0,1)
+                        # t_ijkabc -> Silab
+                        tmp = self.sigma_T2_C(j, k, Wmnie_cc3, t3)
+                        r2[i] += tmp
+                        r2[:,i] += tmp.swapaxes(1,2) 
```

Comment:
My next question is whether it's necessary to handle these terms as separate function calls.  Again, I refer to the comparable code in `t_vikings`, where the contractions appear simpler and thus a bit more readable:

`X1[i] += contract('abc,bc->a',(t3 - t3.swapaxes(0,2)), L[j,k,v,v])`
`X2[i,j] += contract('abc,c->ab',(t3 - t3.swapaxes(0,2)), F[k,v])`
`X2[i,j] += contract('abc,dbc->ad', (2.0*t3 - t3.swapaxes(1,2) - t3.swapaxes(0,2)),ERI[v,k,v,v])`
`X2[i] -= contract('abc,lc->lab', (2.0*t3 - t3.swapaxes(1,2) - t3.swapaxes(0,2)),ERI[j,k,o,v])`